{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HapTree-X Experimental Notebook\n",
    "---\n",
    "\n",
    "## Experimental setup\n",
    "\n",
    "### Data Sources\n",
    "\n",
    "| Sample | Links\n",
    "| --- | ---\n",
    "| **RNA-seq** |\n",
    "| GM12878 (Whole cell) | LongNonPolyA Rep1: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellLongnonpolyaFastqRd1Rep1.fastq.gz) [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellLongnonpolyaFastqRd2Rep1.fastq.gz); LongNonPolyA Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellLongnonpolyaFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellLongnonpolyaFastqRd2Rep2.fastq.gz) \n",
    "| | Pap Rep1: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellPapFastqRd1Rep1.fastq.gz) [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellPapFastqRd2Rep1.fastq.gz); Pap Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellPapFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CellPapFastqRd2Rep2.fastq.gz) \n",
    "| GM12878 (Nucleus) | LongNonPolyA Rep1: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusLongnonpolyaFastqRd1Rep1.fastq.gz) [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusLongnonpolyaFastqRd2Rep1.fastq.gz); LongNonPolyA Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusLongnonpolyaFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusLongnonpolyaFastqRd2Rep2.fastq.gz)\n",
    "| | Pap Rep1: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusPapFastqRd1Rep1.fastq.gz) [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusPapFastqRd2Rep1.fastq.gz); Pap Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusPapFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878NucleusPapFastqRd2Rep2.fastq.gz)\n",
    "| GM12878 (Cytosol) | LongNonPolyA Rep1: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolLongnonpolyaFastqRd1Rep1.fastq.gz) [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolLongnonpolyaFastqRd2Rep1.fastq.gz); LongNonPolyA Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolLongnonpolyaFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolLongnonpolyaFastqRd2Rep2.fastq.gz) \n",
    "| | Pap Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolPapFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolPapFastqRd2Rep2.fastq.gz); Pap Rep2: [FASTQ_1](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolPapFastqRd1Rep2.fastq.gz)  [FASTQ_2](http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/wgEncodeCshlLongRnaSeqGm12878CytosolPapFastqRd2Rep2.fastq.gz) \n",
    "| K562 (Nucleolus) | Total Rep3: [FASTQ_1](http://hgdownload-test.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/releaseLatest/wgEncodeCshlLongRnaSeqK562NucleolusTotalFastqRd1Rep3.fastq.gz) [FASTQ_2](http://hgdownload-test.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCshlLongRnaSeq/releaseLatest/wgEncodeCshlLongRnaSeqK562NucleolusTotalFastqRd2Rep3.fastq.gz) \n",
    "| GIAB samples | [Available here](http://cb.csail.mit.edu/cb/haptreex/datasets/) \n",
    "| **WXS and WGS** |\n",
    "| Whole-exome | [BAM](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA12878/exome_alignment/NA12878.mapped.ILLUMINA.bwa.CEU.exome.20121211.bam) \n",
    "| Whole-genome | [BAM](https://storage.googleapis.com/genomics-public-data/platinum-genomes/bam/NA12877_S1.bam) [VCF](https://storage.googleapis.com/genomics-public-data/platinum-genomes/vcf/NA12878_S1.genome.vcf) \n",
    "| **10X Genomics** |\n",
    "| NA12878 | [FASTQ](http://s3-us-west-2.amazonaws.com/10x.files/samples/genome/2.1.4/NA12878_WGS_v2/NA12878_WGS_v2_fastqs.tar) \n",
    "| NA24385 | [FASTQ](http://s3-us-west-2.amazonaws.com/10x.files/samples/assembly/2.0.0/ash/ash_fastqs.tar) \n",
    "| **Ground truth VCFs** |\n",
    "| NA12878 | [VCF](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/NA12878_HG001/latest/GRCh37/HG001_GRCh37_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_PGandRTGphasetransfer.vcf.gz) \n",
    "| NA24143 | [VCF](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/AshkenazimTrio/HG004_NA24143_mother/latest/GRCh37/HG004_GRCh37_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X_CHROM1-22_v.3.3.2_highconf.vcf.gz) \n",
    "| NA24149 | [VCF](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/AshkenazimTrio/HG003_NA24149_father/latest/GRCh37/HG003_GRCh37_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X_CHROM1-22_v.3.3.2_highconf.vcf.gz) \n",
    "| NA24385 | [VCF](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/AshkenazimTrio/HG002_NA24385_son/latest/GRCh37/HG002_GRCh37_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-22_v.3.3.2_highconf_triophased.vcf.gz) \n",
    "| NA24631 | [VCF](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/ChineseTrio/HG005_NA24631_son/latest/GRCh37/HG005_GRCh37_highconf_CG-IllFB-IllGATKHC-Ion-SOLID_CHROM1-22_v.3.3.2_highconf.vcf.gz) \n",
    "| K562    | [VCF](https://www.encodeproject.org/files/ENCFF752OAX/@@download/ENCFF752OAX.vcf.gz) \n",
    "\n",
    "### Evaluated tools\n",
    "\n",
    "| Tool | Version |\n",
    "|------|-----|\n",
    "| HapTree-X | [v1.0](https://github.com/inumanag/haptreex) |\n",
    "| HapCUT2 | [de33b57](https://github.com/vibansal/HapCUT2/tree/de33b57106898fc46751d55bb338456dfd5dbbe5) |\n",
    "| HapCUT | [0.7 (254c473)](https://github.com/vibansal/hapcut/tree/254c473066565e6b9d11b0e725e17ff910dcfbb1) |\n",
    "| phASER | [1.1.1 (20eb3b)](https://github.com/secastel/phaser/tree/20eb3b44394fe0f2f3e002ffbcf7b0db6ff01966) |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Experimental pipeline\n",
    "\n",
    "### Read alignment and genotyping\n",
    "\n",
    "We used the [following Nextflow script](scripts/gatk-rnaseq.nf) to run the alignment and the best practices RNA-seq pipeline. For exome data, we ran the same pipeline, but have ignored all RNA-seq-related steps (all steps before the `HaplotypeCaller` step).\n",
    "\n",
    "10X samples were aligned as described in [EMA Experimental Notebook](https://github.com/arshajii/ema-paper-data/blob/master/experiments.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phasing\n",
    "\n",
    "#### HapCUT\n",
    "\n",
    "```bash\n",
    "# Extract the HAIRS from a given BAM\n",
    "time -f '%E %Mkb' \\\n",
    "  tools/hapcut/extractHAIRS \\\n",
    "   --bam ${SAMPLE}.bam \\\n",
    "   --vcf ${SAMPLE}.vcf \\\n",
    "   --mbq 0 --mmq 0 \\\n",
    "   --out ${SAMPLE}.frag.hapcut1 \n",
    "\n",
    "# Run phasing\n",
    "time -f '%E %Mkb' \\\n",
    "  tools/hapcut/HAPCUT \\\n",
    "  --fragments ${SAMPLE}.frag.hapcut1 \\\n",
    "  --VCF ${SAMPLE}.vcf \\\n",
    "  --output ${SAMPLE}.hapcut1\n",
    "```\n",
    "\n",
    "For RNA-seq samples, we also added `--maxIS 1000000000 --minIS 0` to each `extractHAIRS` invocation.\n",
    "\n",
    "For joint exome & RNA-seq phasing, we ran HapCUT and HapCUT2 `extractHAIRS` step as follows:\n",
    "```bash\n",
    "# Get RNA-seq HAIRS\n",
    "time -f '%E %Mkb' \\\n",
    "  tools/hapcut/extractHAIRS \\\n",
    "  --bam ${SAMPLE}.bam \\\n",
    "  --vcf ${SAMPLE}.vcf \\\n",
    "  --mbq 0 --mmq 0 --maxIS 1000000000 --minIS 0 \\\n",
    "  --out ${SAMPLE}.rna.frag.hapcut1 \n",
    "\n",
    "# Get exome HAIRS\n",
    "time -f '%E %Mkb' \\\n",
    "  tools/hapcut/extractHAIRS \\\n",
    "  --bam exome.bam \\\n",
    "  --vcf ${SAMPLE}.vcf \\\n",
    "  --mbq 0 --mmq 0 \\\n",
    "  --out ${SAMPLE}.exome.frag.hapcut1  \n",
    "\n",
    "# Concatenate HAIRS\n",
    "cat ${SAMPLE}.rna.frag.hapcut1 ${SAMPLE}.exome.frag.hapcut1 \\\n",
    "  > ${SAMPLE}.frag.hapcut1\n",
    "\n",
    "# Now run HapCUT on the concatenated HAIRS ${SAMPLE}.frag.hapcut1\n",
    "```\n",
    "We used concatenated VCF as described in the HapTree-X phasing subsection.\n",
    "\n",
    "\n",
    "#### HapCUT2\n",
    "\n",
    "```bash\n",
    "# Extract the HAIRS from a given BAM\n",
    "time -f '%E %Mkb' \\\n",
    "  tools/hapcut2/build/extractHAIRS \\\n",
    "   --bam ${SAMPLE}.bam \\\n",
    "   --VCF ${SAMPLE}.vcf \\\n",
    "   --out ${SAMPLE}.frag.hapcut2\n",
    "\n",
    "# Run phasing\n",
    "time -f '%E %Mkb' \\\n",
    "  tools/hapcut2/build/HAPCUT2 \\\n",
    "  --fragments ${SAMPLE}.frag.hapcut2 \\\n",
    "  --VCF ${SAMPLE}.vcf \\\n",
    "  --output ${SAMPLE}.hapcut2\n",
    "```\n",
    "\n",
    "For RNA-seq samples, we also added `--maxIS 1000000000 --minIS 0` to each `extractHAIRS` invocation.\n",
    "\n",
    "For joint exome & RNA-seq phasing, we used the same steps as in HapCUT.\n",
    "\n",
    "\n",
    "#### phASER\n",
    "\n",
    "```bash\n",
    "export PYENV_VERSION=2.7.17\n",
    "time -f '%E %Mkb' \\\n",
    "  python tools/phaser/phaser/phaser.py \\\n",
    "  --vcf ${SAMPLE}.vcf.gz \\\n",
    "  --bam ${SAMPLE}.bam \\\n",
    "  --paired_end 1 --mapq 255 --baseq 0 \\\n",
    "  --sample $1 \\\n",
    "  --o ${SAMPLE}.phaser\n",
    "```\n",
    "\n",
    "For joint exome & RNA-seq phasing, we just changed `--bam` parameter to `--bam ${SAMPLE-DNA}.bam,${SAMPLE-RNA}.bam`.\n",
    "\n",
    "#### HapTree-X\n",
    "\n",
    "WGS, 10X Genomics and exome data was phased via:\n",
    "```bash\n",
    "time -f '%E %Mkb' \\\n",
    "  code/haptreex/haptreex \\\n",
    "  -v ${SAMPLE}.vcf \\\n",
    "  -d ${SAMPLE}.bam \\\n",
    "  -o ${SAMPLE}.haptreex\n",
    "```\n",
    "10X samples were phased with the additional `--10x` parameter.\n",
    "\n",
    "RNA-seq samples were phased via:\n",
    "```bash\n",
    "time -f '%E %Mkb' \\\n",
    "  code/haptreex/haptreex \\\n",
    "  -v ${SAMPLE}.vcf \\\n",
    "  -r ${SAMPLE}.bam \\\n",
    "  -g data/Homo_sapiens.GRCh37.87.gtf \\\n",
    "  -o ${SAMPLE}.haptreex\n",
    "```\n",
    "\n",
    "Joint exome & RNA-seq experiments were ran as:\n",
    "```bash\n",
    "time -f '%E %Mkb' \\\n",
    "  code/haptreex/haptreex \\\n",
    "  -v ${SAMPLE}.vcf \\\n",
    "  -r ${SAMPLE-RNA}.bam \\\n",
    "  -d ${SAMPLE-DNA}.bam \\\n",
    "  -g data/Homo_sapiens.GRCh37.87.gtf \\\n",
    "  -o ${SAMPLE}.haptreex\n",
    "```\n",
    "In this case, each `SAMPLE.vcf` was made by concatenating DNA and RNA VCFs through [merge-wxs.py](scripts/merge-wxs.py): \n",
    "```bash\n",
    "python merge-wxs.py rnaseq.vcf exome.vcf > SAMPLE.vcf\n",
    "```\n",
    "\n",
    "DASE-enabled phasing can be disabled by omitting the `-g data/Homo_sapiens.GRCh37.87.gtf` parameter.\n",
    "Multi-threaded mode can be activated by setting `OMP_NUM_THREADS=N` (N is the number of threads) before running HapTree-X."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the provided [script](scripts/stats.py) to calculate the number of phased SNPs, total phased block lengths, switch error rates and the number of incorrectly phased SNPs:\n",
    "\n",
    "```bash\n",
    "python stats.py truth.vcf output\n",
    "```\n",
    "where `truth.vcf` is a ground-truth phased VCF and `output` is the output of the phaser.\n",
    "\n",
    "Sample output:\n",
    "```\n",
    "$ pypy3 stats.py giab/NA12878/truth-giab.vcf phased_file\n",
    "total_snps,error_snps,total_snps_in_ground_truth,total_edges,total_phased_blocks,switch_error,total_span\n",
    "2799967,155564,964795,2182597,617370,16.1240470773584,234962202\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
